Add dmverity support to the erofs snapshotter using go-dmverity

aadhar-agarwal opened the pull request:
**Add dm-verity support for EROFS layers in containerd**

This PR introduces block-level data integrity verification for EROFS container layers using device-mapper verity (dm-verity):

- Integrates `go-dmverity` to natively create Merkle hash trees and dm-verity devices for EROFS blobs.
  - https://github.com/containerd/go-dmverity
- Adds new configuration: `dmverity_mode` for the snapshotter and `enable_dmverity` for the differ.
- EROFS layers are formatted with a dm-verity hash tree; metadata is stored in `layer.erofs.dmverity` alongside the blob.
- At runtime, each layer is mounted via a dm-verity device (read-only) and verified.
- Mount options for dm-verity are passed via the mount manager and applied per layer.
- Extensive tests cover formatting, mounting, idempotency, error cases, and end-to-end workflow.
- Non-Linux environments gracefully handle unsupported dm-verity operations.

**Note:** When enabled, it requires a Linux kernel with dm-verity support and the `dm_verity` module loaded. If using the veritysetup cli is preferred, please refer to https://github.com/containerd/containerd/pull/12457

```release-note
Add dmverity support to the erofs snapshotter
```

aadhar-agarwal committed:
Add dmverity support to the erofs snapshotter using veritysetup-go

Signed-off-by: Aadhar Agarwal <aadagarwal@microsoft.com>

hsiangkao commented on plugins/snapshots/erofs/erofs.go:
```diff
@@ -108,6 +130,16 @@ func NewSnapshotter(root string, opts ...Opt) (snapshots.Snapshotter, error) {
 		return nil, err
 	}
 
+	if config.enableDmverity {
```
Just my own opinion here,
I guess dm-verity enabling can be on the per snapshot/layer basis?
no need to check dmverity here unless the dmverity is mandatory.

or the boolean should be `dmverity_mode=[auto/off/on]` instead `enable_dmverity` as the differ, because only `force_on` should fail if some layer doesn't enable dmverity.

hsiangkao commented on plugins/snapshots/erofs/erofs.go:
```diff
@@ -200,6 +238,43 @@ func (s *snapshotter) prepareDirectory(ctx context.Context, snapshotDir string,
+// createDmverityErofsMount creates a mount specification for an EROFS layer with dm-verity integrity verification.
+// It reads dm-verity metadata from the .dmverity file and includes all necessary parameters in mount options.
+func (s *snapshotter) createDmverityErofsMount(id string, layerBlob string) (mount.Mount, error) {
+	metadata, err := dmverity.ReadMetadata(layerBlob)
```
I think we could check if `.dmverity` file exists:
if exist, use dmverity unless dmverity_mode is force_off;
if non-exist, error out unless dmverity_mode is force_on;

Also it's possible to support inline dmverity hashtree?
where to check the dmverity signiture?

ChengyuZhu6 replied:
https://github.com/ChengyuZhu6/veritysetup-go/pull/3

hsiangkao commented on plugins/snapshots/erofs/erofs.go:
```diff
@@ -564,6 +636,24 @@ func (s *snapshotter) Remove(ctx context.Context, key string) (err error) {
+			// Close dm-verity device if it exists for this snapshot
+			// The device should already be unmounted by the container runtime before Remove() is called
+			// If the device is still mounted, dmverity.Close will fail safely and we'll log a debug message
+			if s.enableDmverity && id != "" {
```
not quite sure it's clean after mount manager is introduced.

hsiangkao replied:
I think before you make a progress, make sure what is derek's idea on this since if it's a mount manager improvement, that is fine too.

hsiangkao reviewed (commented):
just some premature comments.

aadhar-agarwal commented on docs/snapshotters/erofs.md:
```diff
@@ -178,6 +178,50 @@ introduces additional runtime overhead since all container image reads from
+verity to provide block-level integrity verification for each EROFS layer. This
+method creates a dm-verity device for each layer and mounts it read-only.
+
+Configuration for dm-verity support:
```
Please take a look, just updated

vnmxcmnjane-hub commented:
****

aadhar-agarwal commented on plugins/snapshots/erofs/erofs.go:
```diff
@@ -100,6 +112,16 @@ func NewSnapshotter(root string, opts ...Opt) (snapshots.Snapshotter, error) {
 	}
 
+	// Ensure fsverity and dmverity are not both enabled
+	if config.enableFsverity && config.enableDmverity {
```
When both are enabled fsverity and dmverity are enabled:
- fsverity makes files immutable, but dm-verity requires loop device access that conflicts with this immutability. I see this error:
`$ sudo ./ctr run -rm -t --snapshotter erofs docker.io/library/busybox:latest hello sh
ctr: failed to activate mounts: mount handler failed {erofs /var/lib/containerd/io.containerd.snapshotter.v1.erofs/snapshots/53/layer.erofs  [ro loop]}: failed to open dm-verity device: failed to setup loop device for data: failed to setup loop device for /var/lib/containerd/io.containerd.snapshotter.v1.erofs/snapshots/53/layer.erofs: could not open backing file: /var/lib/containerd/io.containerd.snapshotter.v1.erofs/snapshots/53/layer.erofs: open /var/lib/containerd/io.containerd.snapshotter.v1.erofs/snapshots/53/layer.erofs: operation not permitted`

hsiangkao replied:
> * r.io/library/busybox:latest hello sh ctr: failed to activate mounts: mount handler failed {erofs /var/lib/containerd/io.containerd.snapshotter.v1.erofs/snapshots/53/layer.erofs  [ro loop]}: failed to open dm-verity device: failed to setup loop device for data: failed to setup loop device for /var/lib/containerd/io.containerd.snapshotter.v1.erofs/snapshots/53/layer.erofs: could not open backing file:

I guess that is due to loop device doesn't created with `ro`?

aadhar-agarwal commented on plugins/snapshots/erofs/erofs.go:
```diff
@@ -100,6 +112,16 @@ func NewSnapshotter(root string, opts ...Opt) (snapshots.Snapshotter, error) {
+	}
+
+	// Ensure setImmutable and dmverity are not both enabled
+	if config.setImmutable && config.enableDmverity {
```
Works now, removed the check

aadhar-agarwal commented:
@dmcgowan, @AkihiroSuda, @cpuguy83, @fuweid 

Could you please take a look when you get a chance? :)

ChengyuZhu6 commented on internal/dmverity/dmverity_linux.go:
```diff
@@ -0,0 +1,262 @@
+	defer dmCtrl.Close()
+
+	// Create the device
+	if _, err := dmCtrl.CreateDevice(name); err != nil {
```
How do we ensure loop device is cleaned up when CreateDevice fails?

ChengyuZhu6 reviewed (commented):
@aadhar-agarwal  Good work! Some comments:

ChengyuZhu6 commented on plugins/diff/erofs/dmverity_linux.go:
```diff
@@ -0,0 +1,125 @@
+	// Filesystem sparse allocation makes this efficient
+	if err := os.Truncate(layerBlobPath, fileSize*2); err != nil {
+		return fmt.Errorf("failed to pre-allocate space for hash tree: %w", err)
+	}
```
Allocating 2x the file size wastes space for large image layers. I've exported an API in veritysetup-go to calculate the hash tree storage size, which should be useful here: https://github.com/ChengyuZhu6/veritysetup-go/blob/8e4d0cc1765acdf7a498ff1e7e30a2d82aa07113/pkg/verity/verity.go#L294

aadhar-agarwal commented on plugins/snapshots/erofs/erofs.go:
```diff
@@ -100,6 +111,16 @@ func NewSnapshotter(root string, opts ...Opt) (snapshots.Snapshotter, error) {
+	}
+
+	// Validate dmverityMode
+	if config.dmverityMode != "auto" && config.dmverityMode != "on" && config.dmverityMode != "off" {
```
Added validation when the mode="on"

aadhar-agarwal commented on internal/dmverity/dmverity_linux.go:
```diff
@@ -0,0 +1,316 @@
+//   - No-superblock mode (opts != nil and opts.NoSuperblock == true):
+//     Uses explicitly provided parameters from opts. All dm-verity parameters must be
+//     supplied programmatically since there's no superblock to read from.
+func Open(dataDevice string, name string, hashDevice string, rootHash string, hashOffset uint64, opts *DmverityOptions) (devicePath string, err error) {
```
Updated to use the respective functions from veritysetup-go. Thanks!

hsiangkao commented:
> @aadhar-agarwal Could we trim down the comments to focus on non-obvious logic? Some of them seem to describe what the code already shows clearly.

I tend to avoid trivial comments too (comments almost equal to the code itself), since it cannot bring extra information.
I think comments are used to document design ideas, key workflows and code which is somewhat hard to understand.

dmcgowan commented:
Looking good, I set this to the next milestone so we can get the veritysetup-go library moved into the containerd org as a core dependency.

aadhar-agarwal commented on plugins/diff/erofs/plugin/plugin.go:
```diff
@@ -37,6 +38,10 @@ type Config struct {
 	EnableTarIndex bool `toml:"enable_tar_index"`
+
+	// EnableDmverity enables dm-verity formatting for EROFS layers
+	// Linux only
```
Architecturally, only the mounting requires Linux. The formatting logic is just cryptographic hashing and file I/O.

However, the veritysetup-go library has Linux-specific syscalls (unix.IoctlRetInt, unix.LOOP_CTL_GET_FREE, unix.BLKGETSIZE64) in its pkg/utils package. Even though the actual formatting code (verity.VerityCreate()) doesn't use these, importing the package fails on non-Linux platforms because the syscall constants don't exist. (ci [failure](https://github.com/containerd/containerd/actions/runs/19588200055/job/56101374122))

The dm-verity kernel module is only needed for mounting, not for formatting. I kept the IsSupported() check in the differ plugin to fail early if dm-verity isn't available, but this could be removed to allow formatting on any Linux system

aadhar-agarwal commented on core/mount/mount_linux.go:
```diff
@@ -92,6 +92,17 @@ func (m *Mount) mount(target string) (err error) {
+	// not for the kernel mount syscall. Be specific about which options to filter.
+	filteredOptions := make([]string, 0, len(options))
+	for _, opt := range options {
+		if strings.HasPrefix(opt, "X-containerd.dmverity=") {
```
Sounds good, will create a separate PR for this as it is not needed in this one, and add more background there. Thanks!

aadhar-agarwal replied:
Created https://github.com/containerd/containerd/pull/12557 to handle this, and removed it from this PR

ChengyuZhu6 reviewed (commented):
@aadhar-agarwal Would you mind updating the code with dmverity? To fix linting issues, I have updated some API names in go-dmverity: https://github.com/containerd/go-dmverity/pull/5/files

aadhar-agarwal commented:
> @aadhar-agarwal Would you mind updating the code with dmverity? To fix linting issues, I have updated some API names in go-dmverity: https://github.com/containerd/go-dmverity/pull/5/files

Updated!

ChengyuZhu6 reviewed (commented):
I tested the code, but found the following error when trying to run the container:
```
# ctr run --snapshotter erofs -t ubuntu:22.04 erofs-test sh 
ctr: failed to create erofs mount for parent 1: dm-verity mode is 'on' but .dmverity metadata not found for layer /var/lib/containerd/io.containerd.snapshotter.v1.erofs/snapshots/1/layer.erofs"

```
Did we miss anything?

ChengyuZhu6 reviewed (approved):
LGTM

cpuguy83 commented on internal/dmverity/dmverity_linux.go:
```diff
@@ -0,0 +1,212 @@
+)
+
+func IsSupported() (bool, error) {
+	moduleData, err := os.ReadFile("/proc/modules")
```
Can check /sys/module/dm_verity here so that non-module dm-verity is supported.

aadhar-agarwal commented on plugins/mount/erofs/plugin_linux.go:
```diff
@@ -39,31 +42,106 @@ import (
 
+	// Check if this layer has dm-verity metadata
+	metadata, err := dmverity.ReadMetadata(m.Source)
+	if err == nil && dmverityMode != "off" {
```
Refactored this - Extracted the dm-verity setup into setupDmVerityDevice() with early-return guard clauses, and a small waitForDevice() helper. This is easier to read through now :)
